lib/init.php

```diff
@@ -57,3 +57,122 @@ function gutenberg_menu() {
 	);
 }
 add_action( 'admin_menu', 'gutenberg_menu', 9 );
+
+if ( ! function_exists( 'html_contains_block' ) ) {
+	/**
+	 * Recursively find a block by attribute.
+	 *
+	 * @param  string  $html The html to search in.
+	 * @param  string  $block_name The block type to search for.
+	 * @param  string  $attribute_name The attribute name to search for.
+	 * @param  string  $attribute_value The attribute value to search for.
+	 *
+	 * @return bool    True if block is found, false otherwise
+	 */
+	function html_contains_block( $html, $block_name, $attribute_name = null, $attribute_value = null ) {
+		$at = 0;
+
+		/**
+		 * This is the same regex as the one used in the block parser.
+		 * It is better to use this solution to look for a block's existence
+		 * in a document compared to having to parsing the blocks in the
+		 * document, avoiding all the performance drawbacks of achieving
+		 * a full representation of block content just to check if one block
+		 * is there.
+		 */
+		$pattern = sprintf(
+			'~<!--\s+?wp:%s\s+(?P<attrs>{(?:(?:[^}]+|}+(?=})|(?!}\s+/?-->).)*+)?}\s+)?/?-->~s',
+			// @TODO: we could ensure that there's a namespace on the block name.
+			preg_quote( $block_name, '~' )
+		);
+
+		while ( false !== preg_match( $pattern, $html, $matches, PREG_OFFSET_CAPTURE, $at ) ) {
+
+			// bail if no matches
+			if ( empty ($matches) ) {
+				return false;
+			}
+
+			$at = $matches[0][1] + strlen( $matches[0][0] );
+
+			if ( ! isset( $attribute_name ) ) {
+				return true;
+			}
+
+			// Don't parse JSON if it's not possible for the attribute to exist.
+			if ( ! str_contains( $matches['attrs'][0], "\"{$attribute_name}\":" ) ) {
+				continue;
+			}
+
+			$attrs = json_decode( $matches['attrs'][0], /* as-associative */ true );
+			if ( ! isset( $attrs[ $attribute_name ] ) ) {
+				continue;
+			}
+
+			if ( ! isset( $attribute_value ) ) {
+				return true;
+			}
+
+			if ( $attribute_value === $attrs[ $attribute_name ] ) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+}
+if ( ! function_exists( 'get_template_parts_that_use_menu' ) ) {
+	/**
+	 * Get all template parts that use a menu.
+	 *
+	 * @param int $wp_navigation_id The menu id.
+	 *
+	 * @return array template parts that use the menu
+	 */
+	function get_template_parts_that_use_menu( $wp_navigation_id ) {
+		// get all wp_template_part posts.
+		$wp_template_part_posts = get_posts(
+			array(
+				'post_type'      => 'wp_template_part',
+				'posts_per_page' => -1,
```

Comment:
it looks like we abandoned all limits here, but this is being added to the query for template parts. there could still be a kind of time-bomb in here because some sites will have hundreds if not thousands of template parts, and we're not able to limit our search. this means that for every request to fetch template parts, we will be processing every template part in the database.

it'd be great if we could impose an arbitrary limit or only add this if a specific query arg requests it so that we don't accidentally make the trivial case obstructively slow, when someone doesn't need to know what template parts use the given menu.

if we limit to an arbitrary amount it would seem like we could add some kind of next-page link or cursor id to continue the search on-demand.

Reply:
@dmsnell I thought this was only called to get the field value when the REST query is for `wp_navigation` posts? The `get_callback` invokes this function to get the field value for the `template_parts_that_use_menu` field.

How is it being added to the query for template parts? Maybe I missed something?

If I'm correct, then we could still further optimise this approach by only including the field if certain query params are provided. We don't need it for standard requests. I don't know the best REST etiquette for that but some REST maintainers might be able to assist.

Reply:
> when the REST query is for wp_navigation posts? 

yeah that was my mistake. still, my point is that we're time-bombing the API and I can imagine a handful of normal cases where all we want is the list of `wp_navigation` posts. making the field opt-in would remove that burden and only charge the cost of computation when it's wanted.

this could go _really bad_ on real sites with lots of template parts. say we have a site with a thousand template parts and one navigation menu. even with that single nav menu, we still have to scan all thousand template part posts every time we request the list of nav menus.

Reply:
Let's set a hard limit on the number of items returned in the query. However we want to avoid situations where there are a lot of template parts which causes us to miss the key template parts for the site.

To avoid this we could also consider optimising for common template parts such as `header` and `footer` to ensure these are always checked first. It's likely to serve the 80% use case.

Reply:
> we want to avoid situations where there are a lot of template parts which causes us to miss the key template parts for the site.

doing what we're trying to do here I think is going to require a search index, and one we'll likely need to build and maintain ourselves. I just don't see it being practical to scan the database every time we want to grab `wp_navigation` items.

that can be done, but it's probably easier to do on its own.

> serve the 80% use case

if we _have_ and can _get_ a `header` and `footer` quickly then that would seem like a good idea

Reply:
> to scan the database every time we want to grab wp_navigation items.

I agree which is why I think this comes back around to limiting the scope of this change to it's own endpoint for the block editor only. 

This way we do not register this field on the Posts Controller responses _at all_.

We can easily introduce a new private Core Data selector which queries this endpoint only when necessary. 

This will avoid hitting the DB each time there is a query for `wp_navigation` to Posts Controller. 

If we want to apply further safety limits we can do that _as well_.